src/seccomp/src/compiler.rs

```diff
@@ -0,0 +1,561 @@
+//! `SeccompCondition`, `SeccompAction`, `SeccompCmpOp`, `SeccompCmpArgLen`.
+//!
+//! The compilation goes through a couple of steps, from JSON to BPF:
+//!                  JSON
+//!                   |
+//!            (via serde_json)
+//!                   |
+//!                   V
+//!       collection of `Filter` objects
+//!                   |
+//!         (via Compiler.compile_blob(...))
+//!                   |
+//!                   V
+//!   collection of `SeccompFilter` objects
+//!     (IR - intermediate representation)
+//!                   |
+//!    (via SeccompFilter.into_bpf(...))
+//!                   |
+//!                   V
+//!     collection of `BpfProgram` objects
+
+use super::syscall_table::SyscallTable;
+use seccomp::{
+    BpfThreadMap, Error as SeccompFilterError, SeccompAction, SeccompCondition, SeccompFilter,
+    SeccompRule, SeccompRuleMap,
+};
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::fmt;
+
+/// Errors compiling Filters into BPF
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Erroneous SyscallObject with both `syscall` and `syscalls`
+    MultipleSyscallFields,
+    /// Error from the SeccompFilter
+    SeccompFilter(SeccompFilterError),
+    /// Invalid syscall name for the given arch
+    SyscallName(String, String),
+    /// Erroneous `syscalls` object with argument checks
+    SyscallsWithArgs(Vec<String>),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            MultipleSyscallFields => write!(
+                f,
+                "Cannot use both `syscalls` and `syscall` properties in an object"
+            ),
+            SeccompFilter(ref err) => write!(f, "{}", err),
+            SyscallName(ref syscall_name, ref arch) => write!(
+                f,
+                "Invalid syscall name: {} for given arch: {}",
+                syscall_name, arch
+            ),
+            SyscallsWithArgs(ref syscalls) => write!(
+                f,
+                "The object with the following `syscalls`: {:?} cannot have argument\
+                 conditions. Use the `syscall` property instead or remove the `args` property.",
+                syscalls
+            ),
+        }
+    }
+}
+
+/// Deserializable object encapsulating rules for one or multiple syscalls
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct SyscallObject {
+    /// Name of the syscall
+    syscall: Option<String>,
+    /// Vector of syscall names
+    syscalls: Option<Vec<String>>,
+    /// Optional on-match action
+    action: Option<SeccompAction>,
+    /// Rule conditions
+    #[serde(rename = "args")]
+    conditions: Option<Vec<SeccompCondition>>,
+    /// Unused field, represents a comment property in the JSON format
+    #[allow(dead_code)]
```

Comment:
Since I added the `#[serde(deny_unknown_fields)]` directive on the `SyscallObject` struct, `serde_json` will complain about additional unknown fields. However, we want to allow the optional `comment` property, to specify comments.

Now that I think more about it, it doesn't make sense to store these comments in memory, and we should discard their content after deserialization.

I will introduce a new type, `Comment`, that stores no value and I will implement a custom Deserialize method for it, that simply discards the data.

Reply:
fixed